File: appupdateX/src/main/java/com/azhon/appupdate/dialog/UpdateDialogActivity.java
package com.azhon.appupdate.dialog;

import android.content.Intent;
import android.graphics.drawable.GradientDrawable;
import android.graphics.drawable.StateListDrawable;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.Gravity;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import com.azhon.appupdate.R;
import com.azhon.appupdate.config.UpdateConfiguration;
import com.azhon.appupdate.listener.OnButtonClickListener;
import com.azhon.appupdate.listener.OnDownloadListener;
import com.azhon.appupdate.manager.DownloadManager;
import com.azhon.appupdate.service.DownloadService;
import com.azhon.appupdate.utils.ApkUtil;
import com.azhon.appupdate.utils.Constant;
import com.azhon.appupdate.utils.DensityUtil;
import com.azhon.appupdate.utils.ScreenUtil;

import java.io.File;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

/**
 * 项目名:    AppUpdate
 * 包名       com.azhon.appupdate.dialog
 * 文件名:    UpdateDialogActivity
 * 创建时间:  2022/2/18 on 11:07
 * 描述:     显示升级对话框
 *
 * @author 阿钟
 */


public class UpdateDialogActivity extends AppCompatActivity implements View.OnClickListener, OnDownloadListener {
    private DownloadManager manager;
    private Button update;
    private NumberProgressBar progressBar;
    private OnButtonClickListener buttonClickListener;
    private File apk;
    private int dialogImage, dialogButtonTextColor, dialogButtonColor, dialogProgressBarColor;
    private boolean forcedUpgrade;
    private final int install = 0x45F;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        overridePendingTransition(0, 0);
        setTitle("");
        setContentView(R.layout.dialog_update);
        init();
    }

    /**
     * 初始化布局
     */
    private void init() {
        manager = DownloadManager.getInstance();
        UpdateConfiguration configuration = manager.getConfiguration();
        configuration.setOnDownloadListener(this);
        forcedUpgrade = configuration.isForcedUpgrade();
        buttonClickListener = configuration.getOnButtonClickListener();
        dialogImage = configuration.getDialogImage();
        dialogButtonTextColor = configuration.getDialogButtonTextColor();
        dialogButtonColor = configuration.getDialogButtonColor();
        dialogProgressBarColor = configuration.getDialogProgressBarColor();
        setWindowSize();
        initView();
    }

    private void initView() {
        View ibClose = findViewById(R.id.ib_close);
        ImageView ivBg = findViewById(R.id.iv_bg);
        TextView title = findViewById(R.id.tv_title);
        TextView size = findViewById(R.id.tv_size);
        TextView description = findViewById(R.id.tv_description);
        progressBar = findViewById(R.id.np_bar);
        progressBar.setVisibility(forcedUpgrade ? View.VISIBLE : View.GONE);
        update = findViewById(R.id.btn_update);
        update.setTag(0);
        View line = findViewById(R.id.line);
        update.setOnClickListener(this);
        ibClose.setOnClickListener(this);
        //自定义
        if (dialogImage != -1) {
            ivBg.setBackgroundResource(dialogImage);
        }
        if (dialogButtonTextColor != -1) {
            update.setTextColor(dialogButtonTextColor);
        }
        if (dialogButtonColor != -1) {
            StateListDrawable drawable = new StateListDrawable();
            GradientDrawable colorDrawable = new GradientDrawable();
            colorDrawable.setColor(dialogButtonColor);
            colorDrawable.setCornerRadius(DensityUtil.dip2px(this, 3));
            drawable.addState(new int[]{android.R.attr.state_pressed}, colorDrawable);
            drawable.addState(new int[]{}, colorDrawable);
            update.setBackgroundDrawable(drawable);
        }
        if (dialogProgressBarColor != -1) {
            progressBar.setReachedBarColor(dialogProgressBarColor);
            progressBar.setProgressTextColor(dialogProgressBarColor);
        }
        //强制升级
        if (forcedUpgrade) {
            line.setVisibility(View.GONE);
            ibClose.setVisibility(View.GONE);
        }
        //设置界面数据
        if (!TextUtils.isEmpty(manager.getApkVersionName())) {
            String newVersion = getResources().getString(R.string.dialog_new);
            title.setText(String.format(newVersion, manager.getApkVersionName()));
        }
        if (!TextUtils.isEmpty(manager.getApkSize())) {
            String newVersionSize = getResources().getString(R.string.dialog_new_size);
            size.setText(String.format(newVersionSize, manager.getApkSize()));
            size.setVisibility(View.VISIBLE);
        }
        description.setText(manager.getApkDescription());
    }

    private void setWindowSize() {
        Window dialogWindow = this.getWindow();
        WindowManager.LayoutParams lp = dialogWindow.getAttributes();
        lp.width = (int) (ScreenUtil.getWidth(this) * 0.7f);
        lp.height = WindowManager.LayoutParams.WRAP_CONTENT;
        lp.gravity = Gravity.CENTER;
        dialogWindow.setAttributes(lp);
    }

    @Override
    public void onClick(View v) {
        int id = v.getId();
        if (id == R.id.ib_close) {
            if (!forcedUpgrade) {
                finish();
            }
            //回调点击事件
            if (buttonClickListener != null) {
                buttonClickListener.onButtonClick(OnButtonClickListener.CANCEL);
            }
        } else if (id == R.id.btn_update) {
            if ((int) update.getTag() == install) {
                installApk();
                return;
            }
            if (forcedUpgrade) {
                update.setEnabled(false);
                update.setText(R.string.background_downloading);
            } else {
                finish();
            }
            //回调点击事件
            if (buttonClickListener != null) {
                buttonClickListener.onButtonClick(OnButtonClickListener.UPDATE);
            }
            startService(new Intent(this, DownloadService.class));
        }
    }

    /**
     * 强制更新，点击进行安装
     */
    private void installApk() {
        ApkUtil.installApk(this, Constant.AUTHORITIES, apk);
    }

    @Override
    public void onBackPressed() {
        if (forcedUpgrade) return;
        super.onBackPressed();
    }

    @Override
    public void finish() {
        super.finish();
        overridePendingTransition(0, 0);
    }

    @Override
    public void start() {

    }

    @Override
    public void downloading(int max, int progress) {
        if (max != -1 && progressBar.getVisibility() == View.VISIBLE) {
            int curr = (int) (progress / (double) max * 100.0);
            progressBar.setProgress(curr);
        } else {
            progressBar.setVisibility(View.GONE);
        }
    }

    @Override
    public void done(File apk) {
        this.apk = apk;
        if (forcedUpgrade) {
            update.setTag(install);
            update.setEnabled(true);
            update.setText(R.string.click_hint);
        }
    }

    @Override
    public void cancel() {

    }

    @Override
    public void error(Exception e) {

    }
}

File: appupdate/src/main/java/com/azhon/appupdate/view/UpdateDialogActivity.kt
package com.azhon.appupdate.view

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.drawable.GradientDrawable
import android.graphics.drawable.StateListDrawable
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.Gravity
import android.view.View
import android.view.WindowManager
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import com.azhon.appupdate.R
import com.azhon.appupdate.config.Constant
import com.azhon.appupdate.listener.OnButtonClickListener
import com.azhon.appupdate.listener.OnDownloadListenerAdapter
import com.azhon.appupdate.manager.DownloadManager
import com.azhon.appupdate.service.DownloadService
import com.azhon.appupdate.util.ApkUtil
import com.azhon.appupdate.util.DensityUtil
import com.azhon.appupdate.util.LogUtil
import java.io.File


/**
 * ProjectName: AppUpdate
 * PackageName: com.azhon.appupdate.view
 * FileName:    UpdateDialogActivity
 * CreateDate:  2022/4/7 on 17:40
 * Desc:
 *
 * @author   azhon
 */

class UpdateDialogActivity : AppCompatActivity(), View.OnClickListener {

    private val install = 0x45
    private val error = 0x46
    private val permissionCode = 0x47
    private lateinit var manager: DownloadManager
    private lateinit var apk: File
    private lateinit var progressBar: NumberProgressBar
    private lateinit var btnUpdate: Button

    companion object {
        private const val TAG = "UpdateDialogActivity"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        overridePendingTransition(0, 0)
        manager = try {
            DownloadManager.getInstance()
        } catch (e: IllegalArgumentException) {
            LogUtil.e(TAG, "An exception occurred by DownloadManager=null,please check your code!")
            return
        }
        val layout = manager.config.viewConfig.layoutId
        if (layout != -1) {
            setContentView(layout)
        } else {
            setContentView(R.layout.app_update_dialog_update)
        }
        //system back button
        onBackPressedDispatcher.addCallback(this, object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                backPressed()
            }
        })
        init()
    }

    private fun init() {
        title = ""
        if (manager.config.forcedUpgrade) {
            manager.config.onDownloadListeners.add(listenerAdapter)
        }
        setWindowSize()
        initView(manager)
        if (manager.downloadState) {
            manager.config.onDownloadListeners.add(listenerAdapter)
            btnUpdate.isEnabled = false
            btnUpdate.text = resources.getString(R.string.app_update_background_downloading)
        }
    }

    private fun initView(manager: DownloadManager) {
        val ibClose = findViewById<View>(R.id.app_update_ib_close)
        val vLine = findViewById<View>(R.id.app_update_line)
        val ivBg = findViewById<ImageView>(R.id.app_update_iv_bg)
        val tvTitle = findViewById<TextView>(R.id.app_update_tv_title)
        val tvSize = findViewById<TextView>(R.id.app_update_tv_size)
        val tvDescription = findViewById<TextView>(R.id.app_update_tv_description)
        progressBar = findViewById(R.id.app_update_progress_bar)
        btnUpdate = findViewById(R.id.app_update_btn_update)
        progressBar.visibility = if (manager.config.forcedUpgrade) View.VISIBLE else View.GONE
        btnUpdate.tag = 0
        btnUpdate.setOnClickListener(this)
        ibClose.setOnClickListener(this)
        if (manager.config.viewConfig.dialogImage != -1) {
            ivBg.setBackgroundResource(manager.config.viewConfig.dialogImage)
        }
        if (manager.config.viewConfig.dialogButtonTextColor != -1) {
            btnUpdate.setTextColor(manager.config.viewConfig.dialogButtonTextColor)
        }
        if (manager.config.viewConfig.dialogProgressBarColor != -1) {
            progressBar.reachedBarColor = manager.config.viewConfig.dialogProgressBarColor
            progressBar.setProgressTextColor(manager.config.viewConfig.dialogProgressBarColor)
        }
        if (manager.config.viewConfig.dialogButtonColor != -1) {
            val colorDrawable = GradientDrawable().apply {
                setColor(manager.config.viewConfig.dialogButtonColor)
                cornerRadius = DensityUtil.dip2px(this@UpdateDialogActivity, 3f)
            }
            val drawable = StateListDrawable().apply {
                addState(intArrayOf(android.R.attr.state_pressed), colorDrawable)
                addState(IntArray(0), colorDrawable)
            }
            btnUpdate.background = drawable
        }
        if (manager.config.forcedUpgrade) {
            vLine.visibility = View.GONE
            ibClose.visibility = View.GONE
        }
        if (manager.config.apkVersionName.isNotEmpty()) {
            tvTitle.text =
                String.format(
                    resources.getString(R.string.app_update_dialog_new),
                    manager.config.apkVersionName
                )
        }
        if (manager.config.apkSize.isNotEmpty()) {
            tvSize.text =
                String.format(
                    resources.getString(R.string.app_update_dialog_new_size),
                    manager.config.apkSize
                )
            tvSize.visibility = View.VISIBLE
        }
        tvDescription.text = manager.config.apkDescription
    }

    private fun setWindowSize() {
        val attributes = window.attributes
        attributes.width = DensityUtil.dip2px(this@UpdateDialogActivity, 280f).toInt()
        attributes.height = WindowManager.LayoutParams.WRAP_CONTENT
        attributes.gravity = Gravity.CENTER
        window.attributes = attributes
    }

    override fun onClick(v: View?) {
        when (v?.id) {
            R.id.app_update_ib_close -> {
                if (manager?.config?.forcedUpgrade == false) {
                    finish()
                }
                manager?.config?.onButtonClickListener?.onButtonClick(OnButtonClickListener.CANCEL)
            }

            R.id.app_update_btn_update -> {
                if (btnUpdate.tag == install) {
                    ApkUtil.installApk(this, Constant.AUTHORITIES!!, apk)
                    return
                }
                if (!checkPermission()) {
                    startUpdate()
                }
            }
        }

    }


    /**
     * check Notification runtime permission [DownloadManager.showNotification] is true && when api>=33.
     * @return false: can continue to download, true: request permission.
     */
    private fun checkPermission(): Boolean {
        if (manager?.config?.showNotification == false) {
            LogUtil.d(TAG, "checkPermission: manager.showNotification = false")
            return false
        }
        if (ActivityCompat.checkSelfPermission(
                this, Manifest.permission.POST_NOTIFICATIONS
            ) == PackageManager.PERMISSION_GRANTED
        ) {
            LogUtil.d(TAG, "checkPermission: has permission")
            return false
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            LogUtil.d(TAG, "checkPermission: request permission")
            ActivityCompat.requestPermissions(
                this, arrayOf(Manifest.permission.POST_NOTIFICATIONS), permissionCode
            )
            return true
        }
        return false
    }

    private fun startUpdate() {
        if (manager?.config?.forcedUpgrade == true) {
            btnUpdate.isEnabled = false
            btnUpdate.text = resources.getString(R.string.app_update_background_downloading)
        } else {
            finish()
        }
        manager?.config?.onButtonClickListener?.onButtonClick(OnButtonClickListener.UPDATE)
        startService(Intent(this, DownloadService::class.java))
    }

    private fun backPressed() {
        if (manager?.config?.forcedUpgrade == true) return
        finish()
        manager?.config?.onButtonClickListener?.onButtonClick(OnButtonClickListener.CANCEL)
    }

    override fun finish() {
        super.finish()
        overridePendingTransition(0, 0)
    }

    private val listenerAdapter: OnDownloadListenerAdapter = object : OnDownloadListenerAdapter() {
        override fun start() {
            btnUpdate.isEnabled = false
            btnUpdate.text = resources.getString(R.string.app_update_background_downloading)
        }

        override fun downloading(max: Int, progress: Int) {
            if (max != -1) {
                val curr = (progress / max.toDouble() * 100.0).toInt()
                progressBar.progress = curr
            } else {
                progressBar.visibility = View.GONE
            }
        }

        override fun done(apk: File) {
            this@UpdateDialogActivity.apk = apk
            btnUpdate.tag = install
            btnUpdate.isEnabled = true
            btnUpdate.text = resources.getString(R.string.app_update_click_hint)
        }

        override fun error(e: Throwable) {
            btnUpdate.tag = error
            btnUpdate.isEnabled = true
            btnUpdate.text = resources.getString(R.string.app_update_continue_downloading)
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int, permissions: Array<out String>, grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (permissionCode == requestCode) {
            startUpdate()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        manager?.config?.onDownloadListeners?.remove(listenerAdapter)
    }
}